src/kontrol/kdist/assert.md

```diff
@@ -11,101 +11,196 @@ module KONTROL-ASSERTIONS
+    rule #assert_gt W1 W2 ERR     => #assert (W1 >Int W2)   ERR +Bytes String2Bytes(": " +String Int2String(W1) +String " <= " +String Int2String(W2))
+    rule #assert_lt W1 W2 ERR     => #assert (W1 <Int W2)   ERR +Bytes String2Bytes(": " +String Int2String(W1) +String " >= " +String Int2String(W2))
+    rule #assert_not_eq W1 W2 ERR => #assert (W1 =/=Int W2) ERR +Bytes String2Bytes(": " +String Int2String(W1) +String " == " +String Int2String(W2))
 ```
 
-
 Capturing cheat code calls
 --------------------------
 
 ```k
     rule [cheatcode.call.assertEq]:
-         <k> #cheatcode_call SELECTOR ARGS => #assert_eq #asWord(#range(ARGS, 0, 32)) #asWord(#range(ARGS, 32, 32)) ... </k>
-      requires SELECTOR ==Int selector ( "assertEq(uint256,uint256)" )
+         <k> #cheatcode_call SELECTOR ARGS => #assert_eq #asWord(#range(ARGS, 0, 32)) #asWord(#range(ARGS, 32, 32)) String2Bytes("assertion failed") ... </k>
+      requires SELECTOR ==Int selector ( "assertEq(address,address)" )
         orBool SELECTOR ==Int selector ( "assertEq(bool,bool)"       )
-        orBool SELECTOR ==Int selector ( "assertEq(address,address)" )
         orBool SELECTOR ==Int selector ( "assertEq(bytes32,bytes32)" )
+        orBool SELECTOR ==Int selector ( "assertEq(int256,int256)"   )
+        orBool SELECTOR ==Int selector ( "assertEq(uint256,uint256)" )
+    [preserves-definedness]
+
+    rule [cheatcode.call.assertEq.err]:
+         <k> #cheatcode_call SELECTOR ARGS => #assert_eq #asWord(#range(ARGS, 0, 32)) #asWord(#range(ARGS, 32, 32)) #range(ARGS, 96, #asWord(#range(ARGS, 64, 32))) ... </k>
+      requires SELECTOR ==Int selector ( "assertEq(address,address,string)" )
+        orBool SELECTOR ==Int selector ( "assertEq(bool,bool,string)"       )
+        orBool SELECTOR ==Int selector ( "assertEq(bytes32,bytes32,string)" )
+        orBool SELECTOR ==Int selector ( "assertEq(int256,int256,string)"   )
+        orBool SELECTOR ==Int selector ( "assertEq(uint256,uint256,string)" )
+    [preserves-definedness]
+
+    rule [cheatcode.call.assertNotEq]:
+         <k> #cheatcode_call SELECTOR ARGS => #assert_not_eq #asWord(#range(ARGS, 0, 32)) #asWord(#range(ARGS, 32, 32)) String2Bytes("assertion failed") ... </k>
+      requires SELECTOR ==Int selector ( "assertNotEq(address,address)" )
+        orBool SELECTOR ==Int selector ( "assertNotEq(bool,bool)"       )
+        orBool SELECTOR ==Int selector ( "assertNotEq(bytes32,bytes32)" )
+        orBool SELECTOR ==Int selector ( "assertNotEq(int256,int256)"   )
+        orBool SELECTOR ==Int selector ( "assertNotEq(uint256,uint256)" )
+    [preserves-definedness]
+
+    rule [cheatcode.call.assertNotEq.err]:
+         <k> #cheatcode_call SELECTOR ARGS => #assert_not_eq #asWord(#range(ARGS, 0, 32)) #asWord(#range(ARGS, 32, 32)) #range(ARGS, 96, #asWord(#range(ARGS, 64, 32))) ... </k>
```

Comment:
Hi @anvacaru. Quick question: How is the `96` from `#range(ARGS, 96, #asWord(#range(ARGS, 64, 32)))` derived? Thank you.

Reply:
For any of the `assertNotEq` selectors in the `requires` clause, the first two arguments are static (take `uint256`, for example). This means that each argument will take 32 bytes in the calldata. So `#range(ARGS, 0, 32)` and `#range(ARGS, 32, 32)` handles that.
The third argument, `string`, is dynamic.  To encode it in the calldata, you first have 32 bytes that represent the `length` of the string, and then `length` bytes representing the actual value.
Given, `#range(ARGS, 96, #asWord(#range(ARGS, 64, 32)))`, you first parse `#asWord(#range(ARGS, 64, 32))` to get the length value, and then read that many bytes starting from index `96` of the calldata.

So, 
```
calldata[0] : arg1, 32 bytes
calldata[32]: arg2, 32 bytes
calldata[64]: length of arg3: 32 bytes
calldata[96]: arg3, `length` bytes.

Reply:
Okay. Got it. Thank you!